### PR TITLE
Liq 2.0: add tests for external calls

### PR DESCRIPTION
**Pull request by e18r**
Add tests for `clip.take` using external calls.

* `testFail_not_enough_dai` tests that a user without enough DAI can't complete the call.
* `test_flashsale` tests that a user without enough DAI can complete the call if they use an external call to get DAI.
* `testFail_reentrancy` tests that a user cannot call `clip.take` again from within the `who.clipperCall` function.
* `testFail_redo` tests that a use cannot call `clip.redo` from within the `who.clipperCall` function.

### CODE REVIEW

**Comment by gbalabasquer** [comment]
I think the spacing was already solved in another PR, so this will be probably bringing a lot of conflicts. Can you revert that and leave this pr just changing the main functionality?
If the spacing is not solved 100% with the other pr, we can take care of it later.

File: src/test/clip.t.sol [hunk 1]
```diff
@@ -49,6 +49,28 @@ contract Guy {
     function bark(Dog dog, bytes32 ilk, address urn) external {
         dog.bark(ilk, urn);
     }
+
+    function clipperCall(uint256 owe, uint256 slice, bytes calldata data)
+        external virtual {
+        Vat(address(clip.vat())).suck(address(0), address(this), owe);
```
Comment (gbalabasquer) [inline src/test/clip.t.sol:55]: The test works, but maybe as an actual example I'd go for something closer to the reality where you can actually test that you are receiving the collateral up front.
So we'd need a contract that exchanges your collateral for DAI and you can leave a small profit to show the case of why someone go run it. It'd be a plain arbitrage operation.

File: src/test/clip.t.sol [hunk 2]
```diff
@@ -903,4 +933,36 @@ contract ClipperTest is DSTest {
...
+            id:  1,
+            amt: 25 ether,
+            max: ray(5 ether),
+            who: address(che),
+            data: "hey"
+        });
+        assertEq(vat.dai(che), 0);
+    }
+
+    function testFail_reentrancy() public takeSetup {
```
Comment (gbalabasquer) [inline src/test/clip.t.sol:959]: You could test a second one trying to enter to `redo`.

File: src/test/clip.t.sol (hunk 1 above)
Comment (e18r) [reply src/test/clip.t.sol]: Updated with a more realistic example

File: src/test/clip.t.sol (hunk 2 above)
Comment (e18r) [reply src/test/clip.t.sol:1145]: Added a `redo` test.

File: src/test/clip.t.sol [hunk 3]
```diff
@@ -19,6 +20,82 @@ interface Hevm {
...
+    Exchange uniswap;
+
+    constructor(
+        Clipper clip_,
+        Vat vat_,
+        DSToken gold_,
+        GemJoin goldJoin_,
+        DSToken dai_,
+        DaiJoin daiJoin_,
+        Exchange uniswap_
```
Comment (gbalabasquer) [inline src/test/clip.t.sol:59]: Let's avoid third party names.

Comment (e18r) [reply src/test/clip.t.sol]: done.

**Review by gbalabasquer — APPROVED** [review]
LGTM

**Review by godsflaw — APPROVED** [review]

**Pull request merged**